core/crdt/pncounter.go

```diff
@@ -0,0 +1,213 @@
+)
+
+var (
+	// ensure types implements core interfaces
+	_ core.ReplicatedData = (*PNCounterRegister)(nil)
+	_ core.Delta          = (*PNCounterDelta)(nil)
+)
+
+// PNCounterDelta is a single delta operation for an PNCounterRegister
+type PNCounterDelta struct {
+	DocKey    []byte
+	FieldName string
+	Priority  uint64
+	// SchemaVersionID is the schema version datastore key at the time of commit.
+	//
+	// It can be used to identify the collection datastructure state at the time of commit.
+	SchemaVersionID string
+	Data            []byte
+	FieldValue      client.FieldValue `json:"-"` // should not be marshalled
+}
+
+// GetPriority gets the current priority for this delta.
+func (delta *PNCounterDelta) GetPriority() uint64 {
+	return delta.Priority
+}
+
+// SetPriority will set the priority for this delta.
+func (delta *PNCounterDelta) SetPriority(prio uint64) {
+	delta.Priority = prio
+}
+
+// Marshal encodes the delta using CBOR.
+func (delta *PNCounterDelta) Marshal() ([]byte, error) {
+	b, err := delta.FieldValue.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	delta.Data = b
+
+	h := &codec.CborHandle{}
+	buf := bytes.NewBuffer(nil)
+	enc := codec.NewEncoder(buf, h)
+	err = enc.Encode(delta)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// PNCounterRegister, Last-Writer-Wins Register, is a simple CRDT type that allows set/get
+// of an arbitrary data type that ensures convergence.
+type PNCounterRegister struct {
+	baseCRDT
+}
+
+// NewPNCounterRegister returns a new instance of the PNCounter with the given ID.
+func NewPNCounterRegister(
+	store datastore.DSReaderWriter,
+	schemaVersionKey core.CollectionSchemaVersionKey,
+	key core.DataStoreKey,
+	fieldName string,
+) PNCounterRegister {
+	return PNCounterRegister{newBaseCRDT(store, key, schemaVersionKey, fieldName)}
+}
+
+// Value gets the current register value
+func (reg PNCounterRegister) Value(ctx context.Context) ([]byte, error) {
+	valueK := reg.key.WithValueFlag()
+	buf, err := reg.store.Get(ctx, valueK.ToDS())
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+// Set generates a new delta with the supplied value
+func (reg PNCounterRegister) Set(value client.FieldValue) *PNCounterDelta {
+	return &PNCounterDelta{
+		DocKey:          []byte(reg.key.DocKey),
+		FieldName:       reg.fieldName,
+		FieldValue:      value,
+		SchemaVersionID: reg.schemaVersionKey.SchemaVersionId,
+	}
+}
+
+// Merge implements ReplicatedData interface.
+// It merges two PNCounterRegisty by adding the values together.
+func (reg PNCounterRegister) Merge(ctx context.Context, delta core.Delta) error {
+	d, ok := delta.(*PNCounterDelta)
+	if !ok {
+		return ErrMismatchedMergeType
+	}
+
+	return reg.addValue(ctx, d.FieldValue, d.GetPriority())
+}
+
+func (reg PNCounterRegister) addValue(ctx context.Context, value client.FieldValue, priority uint64) error {
+	var number int64
+	switch v := value.Value().(type) {
+	case int64:
+		number = int64(v)
+	case uint64:
```

Comment:
It because of the way cbor handles mashalling and unmarshalling. I think it uses uint64 for positive numbers and int64 for negatives. I'll double check though as it might not be the case.

Reply:
Was able to remove it ✅